### PR TITLE
ENG-4481 Render the page widgets in parallel (7.1.x)

### PR DESCRIPTION
**Pull request by entgigi**

### CODE REVIEW

File: portal-ui/src/main/java/org/entando/entando/aps/servlet/ControllerServlet.java [hunk 1]
```diff
@@ -195,4 +202,42 @@ protected void outputError(RequestContext reqCtx, HttpServletResponse response)
 	private static final String ATTR_APPLICATION_MODEL = ".freemarker.Application";
 	private static final String ATTR_JSP_TAGLIBS_MODEL = ".freemarker.JspTaglibs";
 
+	public static class EntTemplateLoader implements TemplateLoader {
```
Comment (entgigi) [inline portal-ui/src/main/java/org/entando/entando/aps/servlet/ControllerServlet.java:205]: @eugeniosant i moved from protected to public to make it easy to test, it's ok ?

File: portal-ui/src/main/java/org/entando/entando/aps/system/services/controller/executor/AbstractWidgetExecutorService.java [hunk 2]
```diff
@@ -57,14 +60,35 @@ public abstract class AbstractWidgetExecutorService {
...
 			List<IFrameDecoratorContainer> decorators = this.extractDecorators(reqCtx);
 			Widget[] widgets = page.getWidgets();
-			for (int frame = 0; frame < widgets.length; frame++) {
-				reqCtx.addExtraParam(SystemConstants.EXTRAPAR_CURRENT_FRAME, new Integer(frame));
-				Widget widget = widgets[frame];
-				widgetOutput[frame] = this.buildWidgetOutput(reqCtx, widget, decorators);
+			if (this.parallelWidgetRender) {
+				List<Widget> widgetList = Arrays.asList(widgets);
+				widgetList.parallelStream().forEach(w -> {
+					ReqCtxThreadLocal.init();
```
Comment (entgigi) [inline portal-ui/src/main/java/org/entando/entando/aps/system/services/controller/executor/AbstractWidgetExecutorService.java:76]: i think we should put all the code below in a try cath for throwable and then destroy threadLocal @eugeniosant what do you think ?

File: portal-ui/src/main/java/org/entando/entando/aps/system/services/controller/executor/AbstractWidgetExecutorService.java [hunk 3]
```diff
@@ -57,14 +60,35 @@ public abstract class AbstractWidgetExecutorService {
...
+				ReqCtxThreadLocal.init();
+				for (int frame = 0; frame < widgets.length; frame++) {
+					reqCtx.addExtraParam(SystemConstants.EXTRAPAR_CURRENT_FRAME, frame);
+					Widget widget = widgets[frame];
+					widgetOutput[frame] = this.buildWidgetOutput(reqCtx, widget, decorators);
+				}
+				ReqCtxThreadLocal.destroy();
 			}
 		} catch (Throwable t) {
 			String msg = "Error detected during widget preprocessing";
```
Comment (entgigi) [inline portal-ui/src/main/java/org/entando/entando/aps/system/services/controller/executor/AbstractWidgetExecutorService.java:97]: here i will put a `ReqCtxThreadLocal.init();`  @eugeniosant  what do you think ?

**Review by zonia3000 — APPROVED** [review]

File: portal-ui/src/main/java/org/entando/entando/aps/servlet/ControllerServlet.java (hunk 1 above)
Comment (eugeniosant) [reply portal-ui/src/main/java/org/entando/entando/aps/servlet/ControllerServlet.java:205]: ok

File: portal-ui/src/main/java/org/entando/entando/aps/system/services/controller/executor/AbstractWidgetExecutorService.java (hunk 3 above)
Comment (eugeniosant) [reply portal-ui/src/main/java/org/entando/entando/aps/system/services/controller/executor/AbstractWidgetExecutorService.java:97]: ok

**Pull request merged**